### PR TITLE
856 full index dont fail

### LINKED ISSUE #774: API: Formalise app state
**Is your feature request related to a problem? Please describe.**
In order to show the state of an application (errored/red, caution/yellow, all good/green, etc) the client needs a formalised state.
At the moment, in `v0.0.21`, it looks like the app only has a [string value](https://github.com/epinio/epinio/blob/8101428b71fcc7fae4dac3a66d73024e5f9d238c/internal/api/v1/models/app.go#L19) that could either show a status sentence or number of instances running (haven't played around with this, so could be wrong about specifics). 
- `0/1`
- `Inactive, without workload. Launch via "epinio app push"`

I've tried fetching the stage associated with the app to see if it had anything more, but just get a 404 - `/api/v1/orgs/<org name>/staging/<stage_id>/logs`.

**Describe the solution you'd like**
A clear enum based state regarding the health of the application, a separate state for number of instances running and another for target number of instances.

The enum could be any number of values, they would just need to be defined so we can associate certain behaviour in the UI.

### LINKED ISSUE #856: Don't fail when listing all apps and some namespace goes missing
This is a followup of this: https://github.com/epinio/epinio/pull/837#issuecomment-932082708

It seems that while the test was running, something removed some namespace behind the scenes (or changed the target?) and made the `epinio apps list --all` command fail. We should find what the cause is and make sure this won't fail again.

### PR DESCRIPTION
**Pull request by jimmykarily**
Fixes #856 
Fixes #774

### CODE REVIEW

**Comment by jimmykarily** [comment]
I can't say I'm very satisfied with this solution but I can't come up with something better right now. Also not sure how to test this. I would need to either force the race condition or force an app to get an error some other way.

The `fetch` is the only function that sets the new `Error` field I introduced. If something constructs the app object using another method, this field won't be set. An empty `Error` field, anywhere in the code flow, should mean there is no error. Right now it could also mean, nobody run the function that sets it.
This happens because there is no "standard" way enforced on how to get an app object. You can have an app object that never went through the `fetch` method or one that did. And there is no way to tell unless you follow the code.

Maybe this is ok, but I feel this makes the code hard to follow and could lead to bugs. I'm trying to think where the origin of the problem lies and what needs to be refactored. Any ideas, welcome.

**Comment by jimmykarily** [comment]
I think #774 and #856 could be solved in one go. If there is an error while populating the application's fields, we can just set the application status to "error". In the case of the app being deleted behind the scenes, this "error" state will be transient. Ideally there should be no error and the app should not be returned but delete actions in kubernetes are not instant. We could introduce a "deleting" state but it will be hard to calculate that state.

File: internal/api/v1/stage.go [hunk 1]
```diff
@@ -206,6 +194,11 @@ func (hc ApplicationsController) Stage(w http.ResponseWriter, r *http.Request) A
 		return InternalError(err, "failed to ensure a PersistenVolumeClaim for the application source and cache")
 	}
 
+	cs, err := tekton.NewForConfig(cluster.RestConfig)
+	if err != nil {
+		return InternalError(err, "failed to get access to a tekton client")
+	}
+	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
```
Comment (jimmykarily) [inline internal/api/v1/stage.go:201]: Use this when merged? https://github.com/epinio/epinio/pull/862/files#diff-e02c24d0a09d47ddaca26d2c3540697e3dae4ce4f296b3f7448c6f34dc6dfc12R198

File: internal/application/application.go [hunk 2]
```diff
@@ -93,6 +95,31 @@ func Exists(ctx context.Context, cluster *kubernetes.Cluster, app models.AppRef)
...
 }
 
+// CurrentlyStaging returns true if there is an active (not completed) PipelineRun
+// for this application.
+func CurrentlyStaging(ctx context.Context, cluster *kubernetes.Cluster, namespace, appName string) (bool, error) {
+	cs, err := tekton.NewForConfig(cluster.RestConfig)
+	if err != nil {
+		return false, err
+	}
+	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
```
Comment (jimmykarily) [inline internal/application/application.go:105]: here too? https://github.com/epinio/epinio/pull/862/files#diff-e02c24d0a09d47ddaca26d2c3540697e3dae4ce4f296b3f7448c6f34dc6dfc12R198

**Comment by jimmykarily** [comment]
I opened the PR for a first review. If we like the solution I will add some tests for it (in a very non TDD way).

File: internal/application/application.go [hunk 3]
```diff
@@ -365,3 +395,31 @@ func fetch(ctx context.Context, cluster *kubernetes.Cluster, app *models.App) er
 
 	return nil
 }
+
+// calculateStatus sets the Status field of the App object.
+// To decide what the status value should be it combined various pieces of information.
```
Comment (andreas-kupries) [inline internal/application/application.go:400]:
```suggestion
// To decide what the status value should be it combines various pieces of information.
```

**Review by andreas-kupries — CHANGES_REQUESTED** [review]

File: internal/application/application.go (hunk 3 above)
Comment (manno) [reply internal/application/application.go:400]:
```suggestion
// To decide what the status value should be, it combines various pieces of information.
```

**Review by andreas-kupries — APPROVED** [review]
